File: 61-rotate-list/61-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */ 
    
// 1 -> 2 ->


class Solution {
    int CountNode(ListNode head){
        ListNode curr = head;
        int val = 0;
        while(curr != null){
            curr = curr.next;
            val++;
        }
        return val;
    }
//     i,j 
        
//         i - k
//          j - var
//         last
    
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null) return head;
        int count = CountNode(head);
        k = k%count;
        ListNode slow= head,fast = head;
        int i = 0;
        ListNode lastfast = null;
        while(fast != null){
            if(i > k){
                    slow = slow.next;
            }
            lastfast = fast;
            fast = fast.next;
            i++;
        }
         
        if(slow.next == null){
            return head;
        }
        else {
            lastfast.next = head;
            ListNode t = slow.next;
            slow.next = null;
            return t;
        }
        
    }
}

File: 61-rotate-list/61-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        
        if(head == NULL || head->next == NULL || k == 0) return head;
        
        ListNode *last = head;
        int len = 1;
        
        while(last->next != NULL) {
            last = last->next;
            len++;
        }
        
        k = (k%len);
        // if(k == 0) return head;
        
        ListNode *end = head;
        k = len - k - 1;
        
        while(k--) {
            end = end->next;
        }
        last->next = head;
        ListNode *newHead = end->next;
        end->next = NULL;
        return newHead;
    }
};